### DISCUSSION TITLE
How does history work with multitab terminal emulators

### DISCUSSION THREAD

**Discussion by 0ion9** [fish-shell/fish-shell#8868] [Q&A] [answered]
In theory, all history is logged to `~/.local/share/fish/fish_history`
and loaded when an instance of fish is started.
Therefore, suppose you have your terminal configured to start 3 tabs by default - there should be 3 instances of fish, and each should, at this point, have the same history.

But my actual experience is that, at the time the terminal is started, some of these tabs will have the history I was expecting, and others will have very little history- I'll have to run 'history merge' in order to get the history i was expecting. This is contrary to the docs for history merge, by which you would presume that you only needed to 'history merge' when you want history items that have been added by other fish instances since this instance was started.

In short, this does not work how I expected, and none of the documentation or examining the history file helps me understand why. It's always been confusing enough that I couldn't tell whether it was a bug or just a flaw in my understanding. Can anybody help?

**Comment by ridiculousfish** [comment]
This is  not expected and I have never seen this. One possible explanation is that `time()` is reporting some time in the past for new tabs. Then fish will ignore most history items, as they would appear to be in the future.

Please take a look at the `when:` timestamps in your history file (typically at `~/.local/share/fish/fish_history`). Are the most recent items (last in the file) weirdly in the past? The values are just seconds since epoch (currently 1649730825); if some values are much before that then it would explain this. In particular run a new command in an affected tab and see what value gets output to the file.

If that does not help please file an issue with your system config, and any repro steps you can find, along with whether it repros without third-party customizations, as described by the template. You can use "create issue from discussion" if you like.

**Reply by 0ion9** [reply]
Timestamps appear to be correct. I should also add that tabs that are manually added do not appear to have this problem. The only difference coming to mind is the time gap between creation of fish instances.

It might take some time to figure out means of reproduction. Its unreliability is one of the things that made me suspect race condition/ timing bug.

**Comment by zzhaolei** [comment, marked as answer]
Will this solve your problem temporarily? -> [up-or-search](https://github.com/zzhaolei/up-or-search.fish/blob/main/functions/up-or-search.fish)

**Reply by 0ion9** [reply]
Thanks for the idea. Seems like it should work, albeit i need to amend a different function than up-or-search.
